### PR TITLE
Example Demonstrating Translation for SDG

### PR DESCRIPTION
**Pull request by murthyrudra**
This PR demonstrates how translation can be used in the flow. The `TranslationBlock` uses `FastAPI` endpoint and can support any translation models. We provide scripts to host Facebook' NLLB and IndicTrans v2 models. We also provide script to run the SDG pipeline

## Summary by CodeRabbit

- **New Features**
  - Introduced a comprehensive translation and question-answer generation pipeline for Kannada Wikipedia documents, including translation servers and flow configurations.
  - Added scripts for generating seed data and running the end-to-end pipeline with command-line options.
  - Implemented translation servers using IndicTrans and NLLB models with FastAPI endpoints.
  - Provided a modular translation block for dataset processing.
  - Added a detailed chat prompt template for the Granite 3.3 2B model enhancing AI assistant interactions.

- **Documentation**
  - Added detailed README with step-by-step instructions, pipeline visualization, and sample outputs.
  - Included configuration files for translation, question generation, and response generation.

- **Chores**
  - Added a requirements file specifying necessary Python dependencies.
  - Refined module imports to explicit named imports with controlled public API exports.

### CODE REVIEW

File: examples/translation_example/generate.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+import logging
+from typing import List
+from datasets import Dataset
+from tqdm import tqdm
+
+logger = setup_logger(__name__)
+
+
+### Granite 3.3 2B Chat Template
+@PromptRegistry.register("ibm-granite/granite-3.3-2b-instruct")
```
Comment (abhi1092) [inline examples/translation_example/generate.py]: @murthyrudra, we can move this into 
https://github.com/Red-Hat-AI-Innovation-Team/sdg_hub/blob/main/src/sdg_hub/prompts.py

File: examples/translation_example/indic_trans_server.py [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+        ),
+        "choices": [
+            {
+                "text": translations[0],
+                "index": 0,
+                "logprobs": None,
+                "finish_reason": "stop",
+            }
+        ],
+    }
```
[the remark this answers is not in the document]
Comment (abhi1092) [reply examples/translation_example/indic_trans_server.py]: @murthyrudra this seems like good suggestion for reducing duplicate code

File: examples/translation_example/indic_trans_server.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+        )
+
+        with torch.no_grad():
+            output = indic_en_indic_model.generate(
+                **inputs,
+                use_cache=True,
+                min_length=0,
+                max_length=256,
+                num_beams=5,
+                num_return_sequences=1,
```
[the remark this answers is not in the document]
Comment (abhi1092) [reply examples/translation_example/indic_trans_server.py]: @murthyrudra seems like typo in model name

**Comment by murthyrudra** [comment]
Hi @abhi1092 , I have incorporated the suggestions

File: src/sdg_hub/blocks/__init__.py [hunk 4]
```diff
@@ -1,6 +1,29 @@
 # Local
-from .block import *
-from .filterblock import *
-from .iterblock import *
-from .llmblock import *
-from .utilblocks import *
```
Comment (abhi1092) [inline src/sdg_hub/blocks/__init__.py:6]: @murthyrudra could you revert these changes? Once the #115 is merged you can just rebase it. Right now these are breaking the code

File: examples/translation_example/README.md [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+### Prerequisites
+Before running the pipeline, let us install the required packages.
+```bash
+pip install -r requirements.txt
+```
+
+We need to start two servers one for serving the LLM and the other for serving the translation system. To start the LLM endpoint, run the command in a different terminal
+```bash
+llm serve ibm-granite/granite-3.3-2b-instruct --port 8082 --max_model_len 2048
```
Comment (abhi1092) [inline examples/translation_example/README.md]: Did you mean `vllm serve ibm-granite/granite-3.3-2b-instruct --port 8082 --max_model_len 2048
`  here?

**Review by abhi1092 — CHANGES_REQUESTED** [review]
I tested your PR and it seems to be working. Just couple of things:
* I ran into issues with flash attn when installing `IndicTransToolkit`. It seems to work when I use it without fash-attn
* When generating data, for 100 rows it generated around 52 rows. Seems like the output is not always able to parse/question/response. 
* Seems like the install script here 
https://github.com/AI4Bharat/IndicTrans2/blob/main/huggingface_interface/install.sh
is pointing to python 3.9 and the subsequent steps break. What worked for me was changing it to python>3.10.
* There seems to be some changes made in `src/sdg_hub/blocks/__init__.py` that were breaking. Could you please revert those and rebase the PR if #115  gets merged?

Apart from the above mentioned fix everything else works. Let me know as soon as you push that change we can merge the PR


Examples:
```
Q:
'When did work on the Shivamogga airport resume according to the provided document, and who approved its start?\n\nHere is the document:\n\nWork on the Shivamogga airport began in 2011. Work stopped in 2015 after disagreements with contractors and partners. In 2020, Chief Minister B.S.Yediyurappa approved the start of work. On 28 February 2023, Prime Minister Narendra Modi visited Shivamogga.\n\nAnswer: Work on the Shivamogga airport resumed in 2020, and this move was approved by Chief Minister B.S.Yediyurappa.'
A: 
'The significant event related to the Shivamogga airport that happened in 2023, according to the provided document, was that Prime Minister Narendra Modi visited Shivamogga. However, the document does not mention any specific work commencement or progress regarding the airport project in 2023.'
```
Translated output
```
Q:
'ಒದಗಿಸಲಾದ ದಾಖಲೆಯ ಪ್ರಕಾರ ಶಿವಮೊಗ್ಗ ವಿಮಾನ ನಿಲ್ದಾಣದ ಕೆಲಸ ಯಾವಾಗ ಪುನರಾರಂಭವಾಯಿತು, ಮತ್ತು ಅದರ ಪ್ರಾರಂಭಕ್ಕೆ ಯಾರು ಅನುಮೋದನೆ ನೀಡಿದರು? ದಾಖಲೆಯು ಹೀಗಿದೆಃ ಶಿವಮೊಗ್ಗ ವಿಮಾನ ನಿಲ್ದಾಣದ ಕೆಲಸ 2011 ರಲ್ಲಿ ಪ್ರಾರಂಭವಾಯಿತು. ಗುತ್ತಿಗೆದಾರರು ಮತ್ತು ಪಾಲುದಾರರೊಂದಿಗಿನ ಭಿನ್ನಾಭಿಪ್ರಾಯಗಳ ನಂತರ 2015 ರಲ್ಲಿ ಕೆಲಸ ನಿಂತುಹೋಯಿತು. 2020 ರಲ್ಲಿ, ಮುಖ್ಯಮಂತ್ರಿಗಳು B.S.Yediyurappa ಕೆಲಸವನ್ನು ಪ್ರಾರಂಭಿಸಲು ಅನುಮೋದನೆ ನೀಡಿದರು. 2023 ರ ಫೆಬ್ರವರಿ 28 ರಂದು, ಪ್ರಧಾನಿ ನರೇಂದ್ರ ಮೋದಿ ಅವರು ಶಿವಮೊಗ್ಗಕ್ಕೆ ಭೇಟಿ ನೀಡಿದರು. ಉತ್ತರಃ ಶಿವಮೊಗ್ಗ ವಿಮಾನ ನಿಲ್ದಾಣದ ಕೆಲಸ 2020 ರಲ್ಲಿ ಪುನರಾರಂಭವಾಯಿತು, ಮತ್ತು ಈ ಕ್ರಮವನ್ನು ಮುಖ್ಯಮಂತ್ರಿಗಳು <ID1] ಅನುಮೋದಿಸಿದರು.'
A:
'ಒದಗಿಸಿದ ದಾಖಲೆಯ ಪ್ರಕಾರ, 2023ರಲ್ಲಿ ಸಂಭವಿಸಿದ ಶಿವಮೊಗ್ಗ ವಿಮಾನ ನಿಲ್ದಾಣಕ್ಕೆ ಸಂಬಂಧಿಸಿದ ಮಹತ್ವದ ಘಟನೆಯೆಂದರೆ, ಪ್ರಧಾನಿ ನರೇಂದ್ರ ಮೋದಿ ಅವರು ಶಿವಮೊಗ್ಗಕ್ಕೆ ಭೇಟಿ ನೀಡಿದ್ದರು. ಆದಾಗ್ಯೂ, ದಾಖಲೆಯಲ್ಲಿ 2023ರಲ್ಲಿ ವಿಮಾನ ನಿಲ್ದಾಣ ಯೋಜನೆಗೆ ಸಂಬಂಧಿಸಿದಂತೆ ಯಾವುದೇ ನಿರ್ದಿಷ್ಟ ಕಾರ್ಯದ ಪ್ರಾರಂಭ ಅಥವಾ ಪ್ರಗತಿಯನ್ನು ಉಲ್ಲೇಖಿಸಲಾಗಿಲ್ಲ.'

File: src/sdg_hub/prompts.py [hunk 6]
```diff
@@ -41,3 +42,62 @@ def nemotron_chat_template():
 {%- if add_generation_prompt %}
   {{- '<|start_header_id|>assistant<|end_header_id|>\n\n' }}
 {%- endif %}"""
+
+
+### Granite 3.3 2B Chat Template
+@PromptRegistry.register("ibm-granite/granite-3.3-2b-instruct")
```
Comment (shivchander) [inline src/sdg_hub/prompts.py]:
```suggestion
@PromptRegistry.register("ibm-granite/granite-3.3")
```

If the chat templates for 2b and 8b are the same, could we change this to be common triggered for both the models.

File: examples/translation_example/README.md (hunk 5 above)
Comment (murthyrudra) [reply examples/translation_example/README.md]: yes. it should be `vllm` and not `llm`

Comment (murthyrudra) [reply examples/translation_example/README.md]: > When generating data, for 100 rows it generated around 52 rows. Seems like the output is not always able to parse/question/response.

This issue arises as I am using `granite 3.3 2B Instruct` model which does not generate structured output consistently. If we replace this with a stringer teacher model we shouldn't see any drop

**Comment by murthyrudra** [comment]
Closing this PR in favor of #178
